rosbag2_transport/src/rosbag2_transport/player.cpp

```diff
@@ -901,69 +922,125 @@ void PlayerImpl::load_storage_content()
-        break;
+  rosbag2_storage::SerializedBagMessageSharedPtr message_ptr = take_next_message_from_queue();
+
+  // While we haven't stopped playing
+  while (rclcpp::ok() && !stop_playback_) {
+    // If we had run out of messages before but are starting to play next again, e.g., after a
+    // seek()
+    if (play_next_.load()) {
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
+      skip_message_in_main_play_loop_ = false;
+      cancel_wait_for_next_message_ = false;
+      // Only take again if we need
+      if (nullptr == message_ptr) {
+        message_ptr = take_next_message_from_queue();
       }
     }
-    std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
-    if (rclcpp::ok()) {
-      if (skip_message_in_main_play_loop_) {
-        skip_message_in_main_play_loop_ = false;
-        cancel_wait_for_next_message_ = false;
-        message_ptr = peek_next_message_from_queue();
-        continue;
+
+    // While there's a message to play and we haven't reached the end timestamp yet
+    while (rclcpp::ok() && !stop_playback_ &&
+      message_ptr != nullptr && !shall_stop_at_timestamp(message_ptr->recv_timestamp))
+    {
+      // Sleep until the message's replay time, do not move on until sleep_until returns true
+      // It will always sleep, so this is not a tight busy loop on pause
+      // However, skip sleeping if we're trying to play the next message
+      while (rclcpp::ok() && !stop_playback_ && !play_next_.load() &&
+        !clock_->sleep_until(message_ptr->recv_timestamp))
+      {
+        // Stop sleeping if cancelled
+        if (std::atomic_exchange(&cancel_wait_for_next_message_, false)) {
+          break;
+        }
       }
-      publish_message(message_ptr);
+
+      std::lock_guard<std::mutex> lk(skip_message_in_main_play_loop_mutex_);
+      if (rclcpp::ok()) {
+        // This means that the queue's top() has changed, so we need to update our message
+        if (skip_message_in_main_play_loop_) {
+          skip_message_in_main_play_loop_ = false;
+          cancel_wait_for_next_message_ = false;
+          message_ptr = take_next_message_from_queue();
+          continue;
+        }
+
+        const bool message_published = publish_message(message_ptr);
+        // If we tried to publish because of play_next(), jump the clock
+        if (play_next_.load()) {
+          clock_->jump(message_ptr->recv_timestamp);
+          // If we successfully played next, notify that we're done, otherwise keep trying
+          if (message_published) {
+            play_next_ = false;
+
+            std::lock_guard<std::mutex> lk(finished_play_next_mutex_);
+            finished_play_next_ = true;
+            play_next_result_ = true;
+            finished_play_next_cv_.notify_all();
+          }
+        }
+      }
+      message_ptr = take_next_message_from_queue();
+    }
+
+    // At this point, we're at the end of the queue, there are no more messages to play
```

Comment:
```suggestion
    // At this point, we're at the end of the playback round, there are no more messages to play
```

Reply:
reworded in c6d2ef5951782a6405fc9fbe2d71e769cec7cac6